src/logger.coffee

```diff
@@ -1,156 +1,146 @@
-			nextDelay = Math.max(@logPublishInterval - elapsedTimeMs, 0)
-			Promise.delay(nextDelay)
-		.then =>
-			if !@shouldStop
-				@publishLoop()
-
-	_truncateToMaxLength: (message) =>
-		if message.message.length > @maxLineLength
-			message = _.clone(message)
-			message.message = _.truncate(message.message, { length: @maxLineLength, omission: '[...]' })
-		return message
+ZLIB_TIMEOUT = 100
+COOLDOWN_PERIOD = 5 * 1000
+KEEPALIVE_TIMEOUT = 60 * 1000
+RESPONSE_GRACE_PERIOD = 5 * 1000
+
+MAX_LOG_LENGTH = 10 * 1000
+MAX_PENDING_BYTES = 256 * 1024
+
+class LogBackend
+	constructor: (apiEndpoint, uuid, deviceApiKey) ->
+		@publishEnabled = true
+
+		@_req = null
+		@_dropCount = 0
+		@_writable = true
+
+		@_opts = url.parse("#{apiEndpoint}/device/v2/#{uuid}/log-stream")
+		@_opts.method = 'POST'
+		@_opts.headers = {
+			'Authorization': "Bearer #{deviceApiKey}"
+			'Content-Type': 'application/x-ndjson'
+			'Content-Encoding': 'gzip'
+		}
+
+		# This stream serves serves as a message buffer during reconnections
+		# while we unpipe the old, malfunctioning connection and then repipe a
+		# new one.
+		@_stream = new stream.PassThrough({
+			allowHalfOpen: true
+			# We halve the high watermark because a passthrough stream has two
+			# buffers, one for the writable and one for the readable side. The
+			# write() call only returns false when both buffers are full.
+			highWaterMark: MAX_PENDING_BYTES / 2
+		})
+		@_stream.on 'drain', =>
+			@_writable = true
+			@_flush()
+			if @_dropCount > 0
+				@_write({
+					message: "Warning: Suppressed #{@_dropCount} message(s) due to high load"
+					timestamp: Date.now()
+					isSystem: true
+					isStdErr: true
+				})
+				@_dropCount = 0
+
+	_setup: _.throttle(->
+		@_req = https.request(@_opts)
+
+		# Since we haven't sent the request body yet, and never will,the
+		# only reason for the server to prematurely respond is to
+		# communicate an error. So teardown the connection immediately
+		@_req.on 'response', (res) =>
+			console.log('LogBackend: server responded with status code:', res.statusCode)
+			@_teardown()
+
+		@_req.on('timeout', => @_teardown())
+		@_req.on('close', => @_teardown())
+		@_req.on 'error', (err) =>
+			console.log('LogBackend: unexpected error:', err)
+			@_teardown()
+
+		# Immediately flush the headers. This gives a chance to the server to
+		# respond with potential errors such as 401 authentication error
+		@_req.flushHeaders()
+
+		# We want a very low writable high watermark to prevent having many
+		# chunks stored in the writable queue of @_gzip and have them in
+		# @_stream instead. This is desirable because once @_gzip.flush() is
+		# called it will do all pending writes with that flush flag. This is
+		# now what we want though. If there are 100 items in the queue we want
+		# to write all of them with Z_NO_FLUSH and only afterwards do a
+		# Z_SYNC_FLUSH to maximize compression
+		@_gzip = zlib.createGzip(writableHighWaterMark: 1024)
```

Comment:
What is the unit of the 1024?  It'd be nice to move it to a named constant at the top of the file I think

Reply:
The constant was a separate thing; it allows us to see the key config vars and such at the top of the file without having to dig into where it's actually used, otherwise you could just hardcode all the other constants in-place by the same argument.  The reason for wanting to know the units was so that I could see if it was a sane value (and I should have created separate threads for the two in retrospect)